{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: biopython in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (1.83)\n",
      "Requirement already satisfied: pandas in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (2.1.4)\n",
      "Requirement already satisfied: torch in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (2.1.2)\n",
      "Requirement already satisfied: matplotlib in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (3.8.2)\n",
      "Requirement already satisfied: seaborn in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (0.13.1)\n",
      "Requirement already satisfied: tqdm in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (4.66.1)\n",
      "Requirement already satisfied: numpy in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from biopython) (1.26.3)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from pandas) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from pandas) (2023.3.post1)\n",
      "Requirement already satisfied: tzdata>=2022.1 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from pandas) (2023.4)\n",
      "Requirement already satisfied: filelock in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from torch) (3.13.1)\n",
      "Requirement already satisfied: typing-extensions in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from torch) (4.9.0)\n",
      "Requirement already satisfied: sympy in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from torch) (1.12)\n",
      "Requirement already satisfied: networkx in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from torch) (3.2.1)\n",
      "Requirement already satisfied: jinja2 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from torch) (3.1.3)\n",
      "Requirement already satisfied: fsspec in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from torch) (2023.12.2)\n",
      "Requirement already satisfied: nvidia-cuda-nvrtc-cu12==12.1.105 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from torch) (12.1.105)\n",
      "Requirement already satisfied: nvidia-cuda-runtime-cu12==12.1.105 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from torch) (12.1.105)\n",
      "Requirement already satisfied: nvidia-cuda-cupti-cu12==12.1.105 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from torch) (12.1.105)\n",
      "Requirement already satisfied: nvidia-cudnn-cu12==8.9.2.26 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from torch) (8.9.2.26)\n",
      "Requirement already satisfied: nvidia-cublas-cu12==12.1.3.1 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from torch) (12.1.3.1)\n",
      "Requirement already satisfied: nvidia-cufft-cu12==11.0.2.54 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from torch) (11.0.2.54)\n",
      "Requirement already satisfied: nvidia-curand-cu12==10.3.2.106 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from torch) (10.3.2.106)\n",
      "Requirement already satisfied: nvidia-cusolver-cu12==11.4.5.107 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from torch) (11.4.5.107)\n",
      "Requirement already satisfied: nvidia-cusparse-cu12==12.1.0.106 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from torch) (12.1.0.106)\n",
      "Requirement already satisfied: nvidia-nccl-cu12==2.18.1 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from torch) (2.18.1)\n",
      "Requirement already satisfied: nvidia-nvtx-cu12==12.1.105 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from torch) (12.1.105)\n",
      "Requirement already satisfied: triton==2.1.0 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from torch) (2.1.0)\n",
      "Requirement already satisfied: nvidia-nvjitlink-cu12 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from nvidia-cusolver-cu12==11.4.5.107->torch) (12.3.101)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from matplotlib) (1.2.0)\n",
      "Requirement already satisfied: cycler>=0.10 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from matplotlib) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from matplotlib) (4.47.2)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from matplotlib) (1.4.5)\n",
      "Requirement already satisfied: packaging>=20.0 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from matplotlib) (23.2)\n",
      "Requirement already satisfied: pillow>=8 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from matplotlib) (10.2.0)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from matplotlib) (3.1.1)\n",
      "Requirement already satisfied: importlib-resources>=3.2.0 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from matplotlib) (6.1.1)\n",
      "Requirement already satisfied: zipp>=3.1.0 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from importlib-resources>=3.2.0->matplotlib) (3.17.0)\n",
      "Requirement already satisfied: six>=1.5 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from python-dateutil>=2.8.2->pandas) (1.16.0)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from jinja2->torch) (2.1.3)\n",
      "Requirement already satisfied: mpmath>=0.19 in /home/hap196/.conda/envs/public_epitope/lib/python3.9/site-packages (from sympy->torch) (1.3.0)\n"
     ]
    }
   ],
   "source": [
    "!pip3 install biopython pandas torch matplotlib seaborn tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.PDB import PDBParser\n",
    "import os, sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from collections import defaultdict\n",
    "import torch\n",
    "import warnings\n",
    "import time as time\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "sys.path.append('/n/groups/marks/projects/sequence_structure/scripts')\n",
    "parent_dir = '/n/groups/marks/projects/sequence_structure/sabdab'\n",
    "sys.path.append(parent_dir)\n",
    "from pdb_utils import *\n",
    "from structure_utils import arrange_idx\n",
    "import torch\n",
    "from tqdm import tqdm\n",
    "from tqdm.contrib.concurrent import process_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# antibodies composed of different regions: framework (FR) and complementarity-determining (CDR) regions.\n",
    "# FRs and CDRs are define as sequences of numbers converted to strings\n",
    "\n",
    "# FR1 is defined as positions 1 to 26\n",
    "FR1 = [str(n) for n in list(np.arange(1,26+1))]\n",
    "CDR1 = [str(n) for n in list(np.arange(27,38+1))]\n",
    "FR2 = [str(n) for n in list(np.arange(39,55+1))]\n",
    "CDR2 = [str(n) for n in list(np.arange(56,65+1))]\n",
    "FR3 = [str(n) for n in list(np.arange(66,104+1))]\n",
    "# CDR3 has additional specific positions indicating insertions in the antibody sequence\n",
    "CDR3 = [str(n) for n in list(np.arange(105,117+1))] + [\"111A\",\"111B\",\"111C\",\"111D\",\"111E\",\"111F\",\"112A\",\"112B\",\"112C\",\"112D\",\"112E\",\"112F\",\"112G\"]\n",
    "FR4 = [str(n) for n in list(np.arange(118,129+1))]\n",
    "\n",
    "FRs = FR1 + FR2 + FR3 + FR4\n",
    "CDRs = CDR1 + CDR2 + CDR3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame([[\"FR1\",FR1],\n",
    "                   [\"CDR1\",CDR1],\n",
    "                   [\"FR2\",FR2],\n",
    "                   [\"CDR2\",CDR2],\n",
    "                   [\"FR3\",FR3],\n",
    "                   [\"CDR3\",CDR3],\n",
    "                   [\"FR4\",FR4]],\n",
    "                  columns=[\"ab_loc\",\"imgt_pos\"])\n",
    "\n",
    "# dictionary for quick lookup of region type (FR or CDR) based on a given position\n",
    "df = df.set_index(\"ab_loc\")\n",
    "ab_loc_dict = dict(zip(df[\"imgt_pos\"].explode(),df[\"imgt_pos\"].explode().index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# standardizing the format of unique identifiers in the dataset\n",
    "\n",
    "def fix_uq_id(s):\n",
    "    exp = s.str.split('_', expand=True)\n",
    "    exp_u = exp.iloc[:,1:].apply(lambda x: x.str.upper())\n",
    "    sfix = exp[[0]].join(exp_u)\n",
    "    sfixed = sfix.apply('_'.join, axis=1)\n",
    "    return (sfixed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import and preprocess dataset of antibody interactions, focusing on H chain type ond date 2022_12_21\n",
    "\n",
    "#import the list of interaction summary\n",
    "DATE = '2022_12_21'\n",
    "chain_type = 'H'\n",
    "\n",
    "pdb_repo = os.path.join(parent_dir, f'{DATE}/sabdab_dataset')\n",
    "int_fnm = f'{DATE}/summary_tables/{chain_type}_fil_contact_interactions.csv'\n",
    "int_df = pd.read_csv(os.path.join(parent_dir, int_fnm))\n",
    "int_df['uq_id'] = fix_uq_id(int_df['uq_id'])\n",
    "int_df = int_df.set_index('uq_id')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2088, 14)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# investigating int_df\n",
    "int_df\n",
    "int_df.drop_duplicates().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/n/groups/marks/projects/sequence_structure/sabdab/2022_12_21/sabdab_dataset/7mtb/H(H)_ant(G)_contact_cnts.pt does not have contacts counted\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9cd804eba8894795b34a42b54269f54d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "0it [00:00, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/n/groups/marks/projects/sequence_structure/sabdab/2022_12_21/sabdab_dataset/6pi7/H(F)_ant(D)_contact_cnts.pt does not have contacts counted\n",
      "/n/groups/marks/projects/sequence_structure/sabdab/2022_12_21/sabdab_dataset/3ehb/H(C)_ant(B)_contact_cnts.pt does not have contacts counted\n"
     ]
    }
   ],
   "source": [
    "def _construct_profile(tup):\n",
    "    pdb_repo, pdb, chain, antchain, chain_type = tup\n",
    "    uq_id = f'{pdb}_{chain_type}({chain})_ant({antchain})'\n",
    "    #load the contacts file\n",
    "    fnm = os.path.join(pdb_repo, pdb, f'{chain_type}({chain})_ant({antchain})_contact_cnts.pt')\n",
    "    if os.path.isfile(fnm):\n",
    "        cnts_ld = torch.load(fnm)\n",
    "        #sum each count matrix for each imgt in the count file\n",
    "        cnt_num = {k:np.sum(v) for k,v in cnts_ld.items()}\n",
    "        cnt_num = pd.Series(cnt_num, dtype='float64')\n",
    "        cnt_num['uq_id'] = uq_id\n",
    "        return (cnt_num)\n",
    "    else:\n",
    "        print (f'{fnm} does not have contacts counted')\n",
    "        return (None)\n",
    "    \n",
    "#set up an iterator for easy running\n",
    "N = int_df.shape[0]\n",
    "iterable = zip([pdb_repo]*N, int_df['pdb'], int_df[f'{chain_type}chain'], int_df['antigen_chain'], [chain_type]*N)\n",
    "    \n",
    "construct_list = process_map(_construct_profile, iterable)\n",
    "construct_list = [c for c in construct_list if c is not None]\n",
    "c_df = pd.DataFrame(construct_list).fillna(0).set_index('uq_id')\n",
    "\n",
    "#order it correctly\n",
    "imgt_order, uq_imgt_df, imgt_ii_key, ii_imgt_key = arrange_idx(c_df.columns, giveorder=True)\n",
    "c_df = c_df.loc[:, imgt_order]\n",
    "\n",
    "### Get the depth of each column in the alignment\n",
    "\n",
    "align_df = pd.read_csv(os.path.join(parent_dir, f'{DATE}/summary_tables/summary_{chain_type}_sabdab_alignment.csv'))\n",
    "align_df['uq_id'] = fix_uq_id(align_df['uq_id'])\n",
    "#sub out to the structures that are in the parsed dataset\n",
    "s_align_df = align_df[align_df['uq_id'].isin(int_df.index)]\n",
    "\n",
    "#calculate the depth of the alignment at each column\n",
    "a_df = s_align_df.iloc[:, s_align_df.columns.str.match('\\d+')]\n",
    "align_depth = (a_df != '-').sum(0)\n",
    "#remove cols that we don't have counts for some reason\n",
    "align_depth_k = align_depth[pd.Series(align_depth.index).isin(uq_imgt_df.imgt).values]\n",
    "#normalize depth\n",
    "align_depth_km = align_depth_k / a_df.shape[0]\n",
    "\n",
    "#sub out the matrix for which we have numbering in the sequence alignment\n",
    "indexer = uq_imgt_df.reset_index(drop=False)\n",
    "indexer = indexer.set_index('imgt')\n",
    "keep_idx = indexer.loc[align_depth_k.index, 'index']\n",
    "\n",
    "### Remove columns with low depth\n",
    "\n",
    "d_tr = 0.01 #depth threshold - changed, AH\n",
    "#sub out and normalize the matrix\n",
    "new_idx = keep_idx[align_depth_km > d_tr]\n",
    "cfil_df = c_df.loc[:,new_idx.index]\n",
    "cbin_df = cfil_df.copy()\n",
    "cbin_df[cbin_df > 0] = 1.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#compile the metadata features to look at\n",
    "\n",
    "#import all the metadata for the interactions\n",
    "sum_fnm = f'{DATE}/sabdab_summary_all_{DATE}.tsv'\n",
    "sum_df = pd.read_csv(os.path.join(parent_dir, sum_fnm), sep='\\t')\n",
    "sum_df = sum_df.fillna('')\n",
    "sum_df['uq_id'] = sum_df['pdb'] + '_' + chain_type + '(' + sum_df[f'{chain_type}chain'] + ')_ant(' + sum_df['antigen_chain'] + ')'\n",
    "sum_df = sum_df.set_index('uq_id')\n",
    "sum_df = sum_df[~sum_df.index.duplicated()]\n",
    "\n",
    "#import the antibody types\n",
    "atype_fnm = os.path.join(parent_dir, 'recourses/metadata_files/Atype_summary.tsv')\n",
    "atype_df = pd.read_csv(atype_fnm).fillna('')\n",
    "atype_df['uq_id'] = atype_df['pdb'] + '_' + chain_type + '(' + atype_df[f'{chain_type}chain'] + ')_ant(' + atype_df['antigen_chain'] + ')'\n",
    "atype_df['full_id'] = atype_df['pdb'] + '_' + atype_df['Hchain'] + '_' + atype_df['Lchain'] + '_' + atype_df['antigen_chain']\n",
    "atype_df = atype_df.set_index('uq_id')\n",
    "atype_df = atype_df[~atype_df.index.duplicated()]\n",
    "\n",
    "#compile with the interaction ordering\n",
    "meta_cols = ['antigen_type', 'heavy_species', 'light_species', 'heavy_subclass',\n",
    "             'light_subclass', 'light_ctype']\n",
    "\n",
    "met_df = sum_df.loc[cbin_df.index, meta_cols]\n",
    "met_df['antibody_type'] = atype_df.reindex(met_df.index).loc[:,'antibody_type']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "cbin_met_df = met_df.merge(cbin_df, on=\"uq_id\")\n",
    "cbin_met_df[\"FR_contacts\"] = cbin_met_df[[pos for pos in FRs if pos in cbin_df.columns]].sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.9124260355029585 1.0\n"
     ]
    }
   ],
   "source": [
    "print(cbin_met_df[cbin_met_df[\"antibody_type\"] == \"Fv\"][\"FR_contacts\"].mean(), cbin_met_df[cbin_met_df[\"antibody_type\"] == \"Fv\"][\"FR_contacts\"].median())\n",
    "# print(cbin_met_df[cbin_met_df[\"antibody_type\"] == \"VHH\"][\"FR_contacts\"].mean(), cbin_met_df[cbin_met_df[\"antibody_type\"] == \"VHH\"][\"FR_contacts\"].median())\n",
    "# print(cbin_met_df[cbin_met_df[\"antibody_type\"] == \"scFv\"][\"FR_contacts\"].mean(), cbin_met_df[cbin_met_df[\"antibody_type\"] == \"scFv\"][\"FR_contacts\"].median())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "cbin_df_contact_counts = cbin_df.sum().sort_values(ascending=False)\n",
    "cbin_df_contact_counts = pd.DataFrame(cbin_df_contact_counts, columns=[\"contact_counts\"])\n",
    "\n",
    "cbin_df_contact_counts[\"ab_loc\"] = \"\"\n",
    "cbin_df_contact_counts.loc[cbin_df_contact_counts.index.isin(FR1), \"ab_loc\"] = \"FR1\"\n",
    "cbin_df_contact_counts.loc[cbin_df_contact_counts.index.isin(CDR1), \"ab_loc\"] = \"CDR1\"\n",
    "cbin_df_contact_counts.loc[cbin_df_contact_counts.index.isin(FR2), \"ab_loc\"] = \"FR2\"\n",
    "cbin_df_contact_counts.loc[cbin_df_contact_counts.index.isin(CDR2), \"ab_loc\"] = \"CDR2\"\n",
    "cbin_df_contact_counts.loc[cbin_df_contact_counts.index.isin(FR3), \"ab_loc\"] = \"FR3\"\n",
    "cbin_df_contact_counts.loc[cbin_df_contact_counts.index.isin(CDR3), \"ab_loc\"] = \"CDR3\"\n",
    "cbin_df_contact_counts.loc[cbin_df_contact_counts.index.isin(FR4), \"ab_loc\"] = \"FR4\"\n",
    "\n",
    "cbin_df_contact_counts_noncdr3 = cbin_df_contact_counts[~cbin_df_contact_counts.index.isin(CDR3)]\n",
    "cbin_df_contact_counts_noncdr3_gt_avg = cbin_df_contact_counts_noncdr3[cbin_df_contact_counts_noncdr3[\"contact_counts\"] > cbin_df_contact_counts_noncdr3[\"contact_counts\"].mean()]\n",
    "\n",
    "# cbin_df_contact_counts_noncdr3_gt_avg[cbin_df_contact_counts_noncdr3_gt_avg[\"ab_loc\"].str[:3] != \"CDR\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_resid(tup, imgt_pos):\n",
    "    pdb_repo, pdb, chain, antchain, chain_type = tup\n",
    "    \n",
    "    fnm = os.path.join(pdb_repo, pdb, f'{chain_type}({chain})_ant({antchain})_contact_details.csv')\n",
    "    int_details = pd.read_csv(fnm, index_col=0)\n",
    "    # print(int_details)\n",
    "\n",
    "    imgt_pos_aa = int_details[(int_details[\"PDB\"] == pdb) & (int_details[\"chain_type\"] == chain_type) &\n",
    "                                (int_details[\"antibody_chain\"] == chain) &\n",
    "                                ((int_details[\"antibody_idx\"] == float(imgt_pos)) | (int_details[\"antibody_idx\"] == str(imgt_pos)))][\"antibody_AA\"].unique()\n",
    "\n",
    "    \n",
    "    assert len(imgt_pos_aa) <= 1\n",
    "    if len(imgt_pos_aa) == 1: imgt_pos_aa = imgt_pos_aa[0]\n",
    "    else: imgt_pos_aa = None\n",
    "\n",
    "    return imgt_pos_aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████████████████████████████████████████████████████████████████████| 22/22 [02:40<00:00,  7.31s/it]\n"
     ]
    }
   ],
   "source": [
    "imgt_pos_aas_all = []\n",
    "\n",
    "for imgt_pos in tqdm(cbin_df_contact_counts_noncdr3_gt_avg.index, total=len(cbin_df_contact_counts_noncdr3_gt_avg.index)):\n",
    "\n",
    "    for ind, row in int_df.iterrows():\n",
    "        tup = ('/n/groups/marks/projects/sequence_structure/sabdab/2022_12_21/sabdab_dataset', row[\"pdb\"], row[\"Hchain\"], row[\"antigen_chain\"], chain_type)\n",
    "        pdb_repo, pdb, chain, antchain, chain_type = tup\n",
    "        uq_id = f'{pdb}_{chain_type}({chain})_ant({antchain})'\n",
    "        try:\n",
    "            imgt_pos_aa = get_resid(tup, imgt_pos)\n",
    "            imgt_pos_aas_all.append([imgt_pos, uq_id, imgt_pos_aa])\n",
    "        except:\n",
    "            pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imgt_pos</th>\n",
       "      <th>uq_id</th>\n",
       "      <th>aa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>57</td>\n",
       "      <td>6fe4_H(F)_ant(A)</td>\n",
       "      <td>ASN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>57</td>\n",
       "      <td>7jmo_H(H)_ant(A)</td>\n",
       "      <td>TYR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>57</td>\n",
       "      <td>7cwn_H(K)_ant(B)</td>\n",
       "      <td>SER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>57</td>\n",
       "      <td>1nca_H(H)_ant(N)</td>\n",
       "      <td>ASN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>57</td>\n",
       "      <td>4ypg_H(H)_ant(D)</td>\n",
       "      <td>SER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45834</th>\n",
       "      <td>69</td>\n",
       "      <td>4lsu_H(H)_ant(G)</td>\n",
       "      <td>ARG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45837</th>\n",
       "      <td>69</td>\n",
       "      <td>7vyt_H(H)_ant(T)</td>\n",
       "      <td>GLN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45846</th>\n",
       "      <td>69</td>\n",
       "      <td>5f72_H(S)_ant(K)</td>\n",
       "      <td>LYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45852</th>\n",
       "      <td>69</td>\n",
       "      <td>5umi_H(H)_ant(C)</td>\n",
       "      <td>PRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45854</th>\n",
       "      <td>69</td>\n",
       "      <td>6oge_H(C)_ant(A)</td>\n",
       "      <td>GLN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15714 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      imgt_pos             uq_id   aa\n",
       "0           57  6fe4_H(F)_ant(A)  ASN\n",
       "1           57  7jmo_H(H)_ant(A)  TYR\n",
       "3           57  7cwn_H(K)_ant(B)  SER\n",
       "5           57  1nca_H(H)_ant(N)  ASN\n",
       "6           57  4ypg_H(H)_ant(D)  SER\n",
       "...        ...               ...  ...\n",
       "45834       69  4lsu_H(H)_ant(G)  ARG\n",
       "45837       69  7vyt_H(H)_ant(T)  GLN\n",
       "45846       69  5f72_H(S)_ant(K)  LYS\n",
       "45852       69  5umi_H(H)_ant(C)  PRO\n",
       "45854       69  6oge_H(C)_ant(A)  GLN\n",
       "\n",
       "[15714 rows x 3 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imgt_pos_aas_all_df = pd.DataFrame(imgt_pos_aas_all, columns=[\"imgt_pos\",\"uq_id\",\"aa\"])\n",
    "imgt_pos_aas_all_df[~imgt_pos_aas_all_df[\"aa\"].isnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "imgt_pos\n",
       "57    2085\n",
       "36    2085\n",
       "56    2085\n",
       "67    2085\n",
       "40    2085\n",
       "82    2085\n",
       "27    2085\n",
       "52    2085\n",
       "28    2085\n",
       "65    2085\n",
       "29    2085\n",
       "63    2085\n",
       "55    2085\n",
       "62    2085\n",
       "58    2085\n",
       "35    2085\n",
       "37    2085\n",
       "66    2085\n",
       "38    2085\n",
       "59    2085\n",
       "64    2085\n",
       "69    2085\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imgt_pos_aas_all_df.imgt_pos.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "imgt_pos_w_high_contact_count_conservation = []\n",
    "\n",
    "for imgt_pos in imgt_pos_aas_all_df[\"imgt_pos\"].drop_duplicates():\n",
    "    pos_df = imgt_pos_aas_all_df[imgt_pos_aas_all_df[\"imgt_pos\"] == imgt_pos]\n",
    "\n",
    "    imgt_pos_w_high_contact_count_conservation.append([imgt_pos, cbin_df_contact_counts_noncdr3_gt_avg.loc[imgt_pos][\"ab_loc\"],\n",
    "                                                       cbin_df_contact_counts_noncdr3_gt_avg.loc[imgt_pos][\"contact_counts\"], pos_df[\"aa\"].value_counts(normalize=True).iloc[0],\n",
    "                                                       pos_df[\"aa\"].value_counts(normalize=True).index[0]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "imgt_pos_w_high_contact_count_conservation_df = pd.DataFrame(imgt_pos_w_high_contact_count_conservation, columns=[\"imgt_pos\",\"ab_loc\",\"contact_counts\",\"conservation_highest_pct\",\"conservation_highest_aa\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imgt_pos</th>\n",
       "      <th>ab_loc</th>\n",
       "      <th>contact_counts</th>\n",
       "      <th>conservation_highest_pct</th>\n",
       "      <th>conservation_highest_aa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>57</td>\n",
       "      <td>CDR2</td>\n",
       "      <td>1362.0</td>\n",
       "      <td>0.224595</td>\n",
       "      <td>TYR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>36</td>\n",
       "      <td>CDR1</td>\n",
       "      <td>1351.0</td>\n",
       "      <td>0.380775</td>\n",
       "      <td>SER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>64</td>\n",
       "      <td>CDR2</td>\n",
       "      <td>1333.0</td>\n",
       "      <td>0.206481</td>\n",
       "      <td>SER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>59</td>\n",
       "      <td>CDR2</td>\n",
       "      <td>1310.0</td>\n",
       "      <td>0.225287</td>\n",
       "      <td>SER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>38</td>\n",
       "      <td>CDR1</td>\n",
       "      <td>1135.0</td>\n",
       "      <td>0.281028</td>\n",
       "      <td>TYR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>66</td>\n",
       "      <td>FR3</td>\n",
       "      <td>1120.0</td>\n",
       "      <td>0.345013</td>\n",
       "      <td>TYR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>37</td>\n",
       "      <td>CDR1</td>\n",
       "      <td>1015.0</td>\n",
       "      <td>0.667988</td>\n",
       "      <td>TYR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>35</td>\n",
       "      <td>CDR1</td>\n",
       "      <td>975.0</td>\n",
       "      <td>0.407827</td>\n",
       "      <td>SER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>58</td>\n",
       "      <td>CDR2</td>\n",
       "      <td>879.0</td>\n",
       "      <td>0.232000</td>\n",
       "      <td>SER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>62</td>\n",
       "      <td>CDR2</td>\n",
       "      <td>877.0</td>\n",
       "      <td>0.208238</td>\n",
       "      <td>SER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>55</td>\n",
       "      <td>FR2</td>\n",
       "      <td>777.0</td>\n",
       "      <td>0.179587</td>\n",
       "      <td>TRP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>63</td>\n",
       "      <td>CDR2</td>\n",
       "      <td>684.0</td>\n",
       "      <td>0.583211</td>\n",
       "      <td>GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>29</td>\n",
       "      <td>CDR1</td>\n",
       "      <td>567.0</td>\n",
       "      <td>0.589698</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>65</td>\n",
       "      <td>CDR2</td>\n",
       "      <td>483.0</td>\n",
       "      <td>0.598344</td>\n",
       "      <td>THR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>28</td>\n",
       "      <td>CDR1</td>\n",
       "      <td>388.0</td>\n",
       "      <td>0.335917</td>\n",
       "      <td>PHE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>52</td>\n",
       "      <td>FR2</td>\n",
       "      <td>249.0</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>TRP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>27</td>\n",
       "      <td>CDR1</td>\n",
       "      <td>234.0</td>\n",
       "      <td>0.926724</td>\n",
       "      <td>GLY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>82</td>\n",
       "      <td>FR3</td>\n",
       "      <td>214.0</td>\n",
       "      <td>0.284360</td>\n",
       "      <td>ASN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>40</td>\n",
       "      <td>FR2</td>\n",
       "      <td>214.0</td>\n",
       "      <td>0.386792</td>\n",
       "      <td>HIS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>67</td>\n",
       "      <td>FR3</td>\n",
       "      <td>211.0</td>\n",
       "      <td>0.881517</td>\n",
       "      <td>TYR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>56</td>\n",
       "      <td>CDR2</td>\n",
       "      <td>210.0</td>\n",
       "      <td>0.866029</td>\n",
       "      <td>ILE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>69</td>\n",
       "      <td>FR3</td>\n",
       "      <td>203.0</td>\n",
       "      <td>0.361386</td>\n",
       "      <td>ASP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   imgt_pos ab_loc  contact_counts  conservation_highest_pct  \\\n",
       "0        57   CDR2          1362.0                  0.224595   \n",
       "1        36   CDR1          1351.0                  0.380775   \n",
       "2        64   CDR2          1333.0                  0.206481   \n",
       "3        59   CDR2          1310.0                  0.225287   \n",
       "4        38   CDR1          1135.0                  0.281028   \n",
       "5        66    FR3          1120.0                  0.345013   \n",
       "6        37   CDR1          1015.0                  0.667988   \n",
       "7        35   CDR1           975.0                  0.407827   \n",
       "8        58   CDR2           879.0                  0.232000   \n",
       "9        62   CDR2           877.0                  0.208238   \n",
       "10       55    FR2           777.0                  0.179587   \n",
       "11       63   CDR2           684.0                  0.583211   \n",
       "12       29   CDR1           567.0                  0.589698   \n",
       "13       65   CDR2           483.0                  0.598344   \n",
       "14       28   CDR1           388.0                  0.335917   \n",
       "15       52    FR2           249.0                  0.666667   \n",
       "16       27   CDR1           234.0                  0.926724   \n",
       "17       82    FR3           214.0                  0.284360   \n",
       "18       40    FR2           214.0                  0.386792   \n",
       "19       67    FR3           211.0                  0.881517   \n",
       "20       56   CDR2           210.0                  0.866029   \n",
       "21       69    FR3           203.0                  0.361386   \n",
       "\n",
       "   conservation_highest_aa  \n",
       "0                      TYR  \n",
       "1                      SER  \n",
       "2                      SER  \n",
       "3                      SER  \n",
       "4                      TYR  \n",
       "5                      TYR  \n",
       "6                      TYR  \n",
       "7                      SER  \n",
       "8                      SER  \n",
       "9                      SER  \n",
       "10                     TRP  \n",
       "11                     GLY  \n",
       "12                     THR  \n",
       "13                     THR  \n",
       "14                     PHE  \n",
       "15                     TRP  \n",
       "16                     GLY  \n",
       "17                     ASN  \n",
       "18                     HIS  \n",
       "19                     TYR  \n",
       "20                     ILE  \n",
       "21                     ASP  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imgt_pos_w_high_contact_count_conservation_df.sort_values(by=[\"contact_counts\"], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_contact_info(tup, imgt_pos):\n",
    "    pdb_repo, pdb, chain, antchain, chain_type = tup\n",
    "    \n",
    "    contact_info_file = os.path.join(pdb_repo, pdb, f'{chain_type}({chain})_ant({antchain})_contact_details.csv')\n",
    "    int_details = pd.read_csv(contact_info_file, index_col=0)\n",
    "    int_details[\"ab_aa_idx\"] = int_details[\"antibody_AA\"].astype(str) + int_details[\"antibody_idx\"].astype(str)\n",
    "    int_details[\"ag_aa_idx\"] = int_details[\"antigen_AA\"].astype(str) + int_details[\"antigen_idx\"].astype(str)\n",
    "\n",
    "    ab_aa_idx = int_details[(int_details[\"PDB\"] == pdb) & (int_details[\"chain_type\"] == chain_type) &\n",
    "                        (int_details[\"antibody_chain\"] == chain) &\n",
    "                        ((int_details[\"antibody_idx\"] == float(imgt_pos)) | (int_details[\"antibody_idx\"] == str(imgt_pos)))][\"ab_aa_idx\"].unique()\n",
    "\n",
    "    ag_aa_idx = int_details[(int_details[\"PDB\"] == pdb) & (int_details[\"chain_type\"] == chain_type) &\n",
    "                        (int_details[\"antibody_chain\"] == chain) &\n",
    "                        ((int_details[\"antibody_idx\"] == float(imgt_pos)) | (int_details[\"antibody_idx\"] == str(imgt_pos)))][\"ag_aa_idx\"].to_list()\n",
    "\n",
    "    return ab_aa_idx, ag_aa_idx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████████████████████████████████████████████████████████████████████| 22/22 [04:05<00:00, 11.16s/it]\n"
     ]
    }
   ],
   "source": [
    "ag_contact_aas_all = []\n",
    "\n",
    "for imgt_pos in tqdm(cbin_df_contact_counts_noncdr3_gt_avg.index, total=len(cbin_df_contact_counts_noncdr3_gt_avg.index)):\n",
    "# for imgt_pos in tqdm(cbin_df.columns, total=len(cbin_df.columns)):\n",
    "    \n",
    "    for ind, row in int_df.iterrows():\n",
    "        tup = ('/n/groups/marks/projects/sequence_structure/sabdab/2022_12_21/sabdab_dataset', row[\"pdb\"], row[\"Hchain\"], row[\"antigen_chain\"], chain_type)\n",
    "        pdb_repo, pdb, chain, antchain, chain_type = tup\n",
    "        uq_id = f'{pdb}_{chain_type}({chain})_ant({antchain})'\n",
    "\n",
    "        try:\n",
    "            ab_aa, ag_aa = get_contact_info(tup, imgt_pos)\n",
    "            if len(ab_aa) == 1: ab_aa = ab_aa[0]\n",
    "            ag_contact_aas_all.append([uq_id, ab_aa, ag_aa])\n",
    "        except:\n",
    "            pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "contact_info = pd.DataFrame(ag_contact_aas_all, columns=[\"uq_id\",\"ab_aa_idx\",\"ag_aa_idx\"])\n",
    "contact_info = contact_info[contact_info[\"ag_aa_idx\"].apply(lambda x: len(x)) > 0]\n",
    "contact_info[\"ag_aa_idx\"] = contact_info[\"ag_aa_idx\"].apply(lambda arr: list(arr))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████████████████████████████████████████████████████████████| 17591/17591 [1:29:25<00:00,  3.28it/s]\n"
     ]
    }
   ],
   "source": [
    "contact_info_expl = contact_info.explode(\"ag_aa_idx\")\n",
    "contact_info_expl = contact_info_expl[contact_info_expl[\"uq_id\"].isin(contact_info_expl[\"uq_id\"])]\n",
    "\n",
    "motifs = []\n",
    "for (uq_id1, ag_aa_idx1), grp1 in tqdm(contact_info_expl.groupby([\"uq_id\",\"ag_aa_idx\"]), total=len(contact_info_expl.groupby([\"uq_id\",\"ag_aa_idx\"]))):\n",
    "    ab_motif = grp1[\"ab_aa_idx\"].to_list()\n",
    "    for (uq_id2, ag_aa_idx2), grp2 in contact_info_expl.groupby([\"uq_id\",\"ag_aa_idx\"]):\n",
    "        if (uq_id1 == uq_id2) & (ag_aa_idx1 == ag_aa_idx2): pass\n",
    "        else:\n",
    "            if ag_aa_idx1[:3] == ag_aa_idx2[:3]:\n",
    "                ab_motif_intersection = set(ab_motif).intersection(set(grp2[\"ab_aa_idx\"].to_list()))\n",
    "                if len(ab_motif_intersection) > 1:\n",
    "                    motifs.append([uq_id1, ag_aa_idx1[:3], ab_motif_intersection])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "mdf = pd.DataFrame(motifs, columns=[\"uq_id\",\"ag_aa\",\"ab_motif\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "mdf.to_csv(\"all_motifs\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mdf = pd.DataFrame(motifs, columns=[\"uq_id\",\"ag_aa\",\"ab_motif\"])\n",
    "mdf[\"ab_motif\"] = mdf[\"ab_motif\"].apply(lambda s: sorted(list(s)))#.astype(str)\n",
    "mdf\n",
    "mdf.to_csv(\"all_motifs\")\n",
    "# mdf.sort_values(by=[\"ag_aa\",\"ab_motif\"])\n",
    "\n",
    "# for abm, grp in mdf.groupby(\"ab_motif\"):\n",
    "#     print(grp)\n",
    "\n",
    "# # mdf.value_counts()\n",
    "# mdf[mdf.duplicated(keep=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uq_id</th>\n",
       "      <th>ab_aa_idx</th>\n",
       "      <th>ag_aa_idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6fe4_H(F)_ant(A)</td>\n",
       "      <td>ASN57</td>\n",
       "      <td>[ASP35, ASP36]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7jmo_H(H)_ant(A)</td>\n",
       "      <td>TYR57</td>\n",
       "      <td>[GLY416, LYS417, ASP420, TYR421]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2086</th>\n",
       "      <td>7jmo_H(H)_ant(A)</td>\n",
       "      <td>SER36</td>\n",
       "      <td>[LYS458, TYR473, GLN474, ALA475]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4170</th>\n",
       "      <td>6fe4_H(F)_ant(A)</td>\n",
       "      <td>GLY64</td>\n",
       "      <td>[ASP35]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4171</th>\n",
       "      <td>7jmo_H(H)_ant(A)</td>\n",
       "      <td>SER64</td>\n",
       "      <td>[THR415, GLY416, ASP420, ASN460]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6256</th>\n",
       "      <td>7jmo_H(H)_ant(A)</td>\n",
       "      <td>GLY59</td>\n",
       "      <td>[TYR421, ARG457, LYS458, SER459, ASN460]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8340</th>\n",
       "      <td>6fe4_H(F)_ant(A)</td>\n",
       "      <td>TYR38</td>\n",
       "      <td>[ASP35, THR37, SER50]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8341</th>\n",
       "      <td>7jmo_H(H)_ant(A)</td>\n",
       "      <td>TYR38</td>\n",
       "      <td>[LYS417, LEU455, PHE456]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10425</th>\n",
       "      <td>6fe4_H(F)_ant(A)</td>\n",
       "      <td>ARG66</td>\n",
       "      <td>[ASN33, ASP35, THR37, TRP48]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10426</th>\n",
       "      <td>7jmo_H(H)_ant(A)</td>\n",
       "      <td>PHE66</td>\n",
       "      <td>[THR415, GLY416]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12511</th>\n",
       "      <td>7jmo_H(H)_ant(A)</td>\n",
       "      <td>ASN37</td>\n",
       "      <td>[TYR473, ALA475]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14596</th>\n",
       "      <td>7jmo_H(H)_ant(A)</td>\n",
       "      <td>SER35</td>\n",
       "      <td>[LYS458]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16681</th>\n",
       "      <td>7jmo_H(H)_ant(A)</td>\n",
       "      <td>SER58</td>\n",
       "      <td>[TYR421, ARG457, LYS458, SER459, TYR473]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22936</th>\n",
       "      <td>7jmo_H(H)_ant(A)</td>\n",
       "      <td>GLY63</td>\n",
       "      <td>[TYR421, ASN460]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25021</th>\n",
       "      <td>7jmo_H(H)_ant(A)</td>\n",
       "      <td>THR29</td>\n",
       "      <td>[ALA475, GLY476, SER477]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27105</th>\n",
       "      <td>6fe4_H(F)_ant(A)</td>\n",
       "      <td>THR65</td>\n",
       "      <td>[ASP35]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29191</th>\n",
       "      <td>7jmo_H(H)_ant(A)</td>\n",
       "      <td>PHE28</td>\n",
       "      <td>[ALA475, GLY476, ASN487]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31275</th>\n",
       "      <td>6fe4_H(F)_ant(A)</td>\n",
       "      <td>TRP52</td>\n",
       "      <td>[SER77, GLY78, SER79]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33361</th>\n",
       "      <td>7jmo_H(H)_ant(A)</td>\n",
       "      <td>GLY27</td>\n",
       "      <td>[GLY476, PHE486, ASN487]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  uq_id ab_aa_idx                                 ag_aa_idx\n",
       "0      6fe4_H(F)_ant(A)     ASN57                            [ASP35, ASP36]\n",
       "1      7jmo_H(H)_ant(A)     TYR57          [GLY416, LYS417, ASP420, TYR421]\n",
       "2086   7jmo_H(H)_ant(A)     SER36          [LYS458, TYR473, GLN474, ALA475]\n",
       "4170   6fe4_H(F)_ant(A)     GLY64                                   [ASP35]\n",
       "4171   7jmo_H(H)_ant(A)     SER64          [THR415, GLY416, ASP420, ASN460]\n",
       "6256   7jmo_H(H)_ant(A)     GLY59  [TYR421, ARG457, LYS458, SER459, ASN460]\n",
       "8340   6fe4_H(F)_ant(A)     TYR38                     [ASP35, THR37, SER50]\n",
       "8341   7jmo_H(H)_ant(A)     TYR38                  [LYS417, LEU455, PHE456]\n",
       "10425  6fe4_H(F)_ant(A)     ARG66              [ASN33, ASP35, THR37, TRP48]\n",
       "10426  7jmo_H(H)_ant(A)     PHE66                          [THR415, GLY416]\n",
       "12511  7jmo_H(H)_ant(A)     ASN37                          [TYR473, ALA475]\n",
       "14596  7jmo_H(H)_ant(A)     SER35                                  [LYS458]\n",
       "16681  7jmo_H(H)_ant(A)     SER58  [TYR421, ARG457, LYS458, SER459, TYR473]\n",
       "22936  7jmo_H(H)_ant(A)     GLY63                          [TYR421, ASN460]\n",
       "25021  7jmo_H(H)_ant(A)     THR29                  [ALA475, GLY476, SER477]\n",
       "27105  6fe4_H(F)_ant(A)     THR65                                   [ASP35]\n",
       "29191  7jmo_H(H)_ant(A)     PHE28                  [ALA475, GLY476, ASN487]\n",
       "31275  6fe4_H(F)_ant(A)     TRP52                     [SER77, GLY78, SER79]\n",
       "33361  7jmo_H(H)_ant(A)     GLY27                  [GLY476, PHE486, ASN487]"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contact_info[contact_info[\"uq_id\"].isin(contact_info[\"uq_id\"].unique()[:2])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "contact_info.to_csv(\"all_contact_info\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
